packages/service-library/src/servicelib/progress_bar.py

```diff
@@ -70,51 +100,76 @@ async def __aexit__(self, exc_type, exc_value, traceback) -> None:
 
     async def _update_parent(self, value: float) -> None:
         if self._parent:
-            await self._parent.update(value / self.steps)
+            await self._parent.update(value)
 
-    async def _report_external(self, value: float, force: bool = False) -> None:
+    async def _report_external(self, value: float, *, force: bool = False) -> None:
         if not self.progress_report_cb:
             return
 
         with log_catch(logger, reraise=False):
             # NOTE: only report if at least a percent was increased
             if (force and value != self._last_report_value) or (
-                ((value - self._last_report_value) / self.steps) > 0.01
+                (value - self._last_report_value) > _MIN_PROGRESS_UPDATE_PERCENT
             ):
-                await self.progress_report_cb(value / self.steps)
+                call = self.progress_report_cb(value)
+                if isawaitable(call):
+                    await call
                 self._last_report_value = value
 
     async def start(self) -> None:
-        await self._report_external(0, force=True)
-
-    async def update(self, value: float = 1) -> None:
+        await self.set_(0)
+
+    def _compute_progress(self, steps: float) -> float:
+        if not self.step_weights:
+            return steps / self.num_steps
+        weight_index = int(steps)
+        return (
+            sum(self.step_weights[:weight_index])
+            + steps % 1 * self.step_weights[weight_index]
+        )
+
+    async def update(self, steps: float = 1) -> None:
+        parent_update_value = 0.0
         async with self._continuous_value_lock:
-            new_progress_value = self._continuous_progress_value + value
-            if new_progress_value > self.steps:
-                new_progress_value = round(new_progress_value)
-            if new_progress_value > self.steps:
+            new_steps_value = self._current_steps + steps
+            if new_steps_value > self.num_steps:
```

Comment:
`if new_steps_value > self.num_steps` is repeated twice
is there a reason to round only if you meet that condition or you can merge both if statements?

Reply:
yes, I had to think a gain, but the reason why is:
- first there is a check (since it is a float value it can be slightly above the max value), therefore it is rounded
- if it is still above the value then comes the warning